csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedElement.qll

```diff
@@ -0,0 +1,563 @@
+private import TranslatedStmt
+private import IRConstruction
+private import semmle.code.csharp.ir.Util
+private import semmle.code.csharp.ir.internal.IRCSharpLanguage as Language
+
+/**
+ * Gets the built-in `int` type.
+ */
+IntType getIntType() { any() }
+
+ArrayType getArrayOfDim(int dim, Type type) {
+  result.getRank() = dim and
+  result.getElementType() = type
+}
+
+/**
+ * Gets the "real" parent of `expr`. This predicate treats conversions as if
+ * they were explicit nodes in the expression tree, rather than as implicit
+ * nodes as in the regular AST representation.
+ */
+private Element getRealParent(Expr expr) { result = expr.getParent() }
+
+/**
+ * Holds if `expr` is a constant of a type that can be replaced directly with
+ * its value in the IR. This does not include address constants as we have no
+ * means to express those as QL values.
+ */
+predicate isIRConstant(Expr expr) { exists(expr.getValue()) }
+
+// Pulled out to work around QL-796
+private predicate isOrphan(Expr expr) { not exists(getRealParent(expr)) }
+
+/**
+ * Holds if `expr` should be ignored for the purposes of IR generation due to
+ * some property of `expr` or one of its ancestors.
+ */
+private predicate ignoreExprAndDescendants(Expr expr) {
+  // Ignore parentless expressions
+  isOrphan(expr)
+  or
+  // Ignore the constants in SwitchCase, since their values are embedded in the
+  // CaseEdge.
+  getRealParent(expr) instanceof CaseStmt
+  or
+  // Ignore descendants of constant expressions, since we'll just substitute the
+  // constant value.
+  isIRConstant(getRealParent(expr))
+  or
+  // The `DestructorCall` node for a `DestructorFieldDestruction` has a `FieldAccess`
+  // node as its qualifier, but that `FieldAccess` does not have a child of its own.
+  // We'll ignore that `FieldAccess`, and supply the receiver as part of the calling
+  // context, much like we do with constructor calls.
+  // TODO: Deal with C# finalizers and Dispose methods at some point
+  //  expr.getParent().(DestructorCall).getParent() instanceof DestructorFieldDestruction or
+  //  exists(NewArrayExpr newExpr |
+  //    // REVIEW: Ignore initializers for `NewArrayExpr` until we determine how to
+  //    // represent them.
+  //    newExpr.getInitializer().getFullyConverted() = expr
+  //  ) or
+  ignoreExprAndDescendants(getRealParent(expr)) // recursive case
+}
+
+/**
+ * Holds if `expr` (not including its descendants) should be ignored for the
+ * purposes of IR generation.
+ */
+private predicate ignoreExprOnly(Expr expr) {
+  //  exists(NewOrNewArrayExpr newExpr |
+  //    // Ignore the allocator call, because we always synthesize it. Don't ignore
+  //    // its arguments, though, because we use them as part of the synthesis.
+  //    newExpr.getAllocatorCall() = expr
+  //  ) or
+  not translateFunction(expr.getEnclosingCallable())
+  or
+  // Ignore size of arrays when translating
+  (expr.getParent() instanceof ArrayCreation and expr.hasValue()) or
```

Comment:
I'm ignoring it since it is not used during instruction generation (as an immediate value or operand). Should I use it?

Reply:
It might be something to evaluate, e.g.
```
int[] items = new int[Foo()];
```
In this case, we'd want to translate `Foo()`. So in the general case, you'd want to evaluate the expression and have the result in an instruction.